### PR TITLE
Fixed registration of pthread_atfork handlers (876)

### LINKED ISSUE #876: Impossible to do a composer install when elastic apm is enabled
**Describe the bug**
Impossible to do a composer install when elastic apm is enabled

**To Reproduce**
Use docker php image : docker run -it --rm php:8.1.16-bullseye bash
In docker container :
```
curl -sS https://getcomposer.org/installer | php -- --install-dir=/usr/local/bin --filename=composer
apt-get update
apt-get install zip procps
composer require --dev phpunit/phpunit 

# This install is ok
rm -rf vendor/ ~/.composer/cache/
composer install 

# Install APM
curl -sSLO https://github.com/elastic/apm-agent-php/releases/download/v1.8.0/apm-agent-php_1.8.0_all.deb
dpkg -i apm-agent-php_1.8.0_all.deb

# This install is ko, php keep running
rm -rf vendor/ ~/.composer/cache/
composer install
```

By running a `ps auxf` in another terminal, we ca see a lot of php process pending
```
root@2689d3ad1019:/# ps auxf
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root         816  2.0  0.0   4164  3356 pts/1    Ss   11:37   0:00 bash
root         822  0.0  0.0   6760  2864 pts/1    R+   11:37   0:00  \_ ps auxf
root           1  0.0  0.0   4164  2328 pts/0    Ss   11:22   0:00 bash
root         778 90.1  0.3 667016 56808 pts/0    R+   11:36   0:11 php /usr/local/bin/composer install
root         806  0.0  0.2 667016 36004 pts/0    S+   11:36   0:00  \_ php /usr/local/bin/composer install
root         807  0.0  0.2 667016 36004 pts/0    S+   11:36   0:00  \_ php /usr/local/bin/composer install
root         808  0.0  0.2 667016 36008 pts/0    S+   11:36   0:00  \_ php /usr/local/bin/composer install
root         809  0.0  0.2 667016 36008 pts/0    S+   11:36   0:00  \_ php /usr/local/bin/composer install
root         810  0.0  0.2 667016 36008 pts/0    S+   11:36   0:00  \_ php /usr/local/bin/composer install
root         811  0.0  0.2 667016 36008 pts/0    S+   11:36   0:00  \_ php /usr/local/bin/composer install
root         812  0.0  0.2 667016 36008 pts/0    S+   11:36   0:00  \_ php /usr/local/bin/composer install
root         813  0.0  0.2 667016 36008 pts/0    S+   11:36   0:00  \_ php /usr/local/bin/composer install
root         814  0.0  0.2 667016 36008 pts/0    S+   11:36   0:00  \_ php /usr/local/bin/composer install
root         815  0.0  0.2 667016 36008 pts/0    S+   11:36   0:00  \_ php /usr/local/bin/composer install
```

I tried to install older version of apm, 1.7.2 is broken too, 1.7.1 is working...
It's important to remove composer vendor/cache before install because it's vendor downloading that generate this behavior :)

**Expected behavior**
Composer install should work :)

### LINKED ISSUE #962: Child processes spawned from PHP processes may deadlock on Debian Bullseye
**Describe the bug**
When the Elastic APM Agent for PHP is enabled, child processes spawned from PHP processes deadlock after the fork when the parent (PHP) process has called `pthread_create`. This seems to be due to a bug in glibc fixed in the following commit: https://sourceware.org/git/?p=glibc.git;a=commit;h=52a103e237329b9f88a28513fe7506ffc3bd8ced. Unfortunately, Debian Bullseye bundles an older version of glibc. I think the change that introduced the issue is part of #857. This is probably also the underlying cause of #876.

I understand that there might not be much the Elastic APM Agent for PHP can do for that, since it is due to Debian Bullseye bundling an older version of glibc. However, I'm posting this in the hope it can help other people experiencing the issue and that we can find a workaround for that bug.

Additional details about my investigation are provided in the Dockerfile code snippet below.

**To Reproduce**
1. Build the following Dockerfile:
```Dockerfile
FROM php:8.0-cli

RUN curl -fsSL https://github.com/elastic/apm-agent-php/releases/download/v1.8.3/apm-agent-php_1.8.3_all.deb > /tmp/apm-agent.deb \
    && dpkg --install /tmp/apm-agent.deb

# This triggers the bug fixed in the following glibc commit: https://sourceware.org/git/?p=glibc.git;a=commit;h=52a103e237329b9f88a28513fe7506ffc3bd8ced.
# Debian Bullseye uses glibc 2.31, which is an older release.
# The Elastic APM Agent for PHP calls pthread_atfork from a pthread_atfork callback here: https://github.com/elastic/apm-agent-php/blob/da2e59b9cd0acc58fcec067d5e282d37261dc56b/src/ext/platform_threads_linux.c#L532. This triggers the deadlock.
# The pthread_atfork handlers are called from __run_fork_handlers, which is itself called from https://github.com/bminor/glibc/blob/9ea3686266dca3f004ba874745a4087a89682617/sysdeps/nptl/fork.c#L56-L58.
# In order to trigger the bug, the process must have called pthread_create at least once, so that the multiple_threads flag is set.
# cURL calls pthread_create, which sets the multiple_threads flag: https://github.com/bminor/glibc/blob/9ea3686266dca3f004ba874745a4087a89682617/nptl/descr.h#L131-L146.
# This in turn causes __run_fork_handlers to acquire atfork_lock: https://github.com/bminor/glibc/blob/9ea3686266dca3f004ba874745a4087a89682617/nptl/register-atfork.c#L110-L117.
# When pthread_atfork tries to acquire atfork_lock, it blocks because it's already held by __run_fork_handlers: https://github.com/bminor/glibc/blob/9ea3686266dca3f004ba874745a4087a89682617/nptl/register-atfork.c#L40.
RUN echo '<?php\n\
$ch = curl_init("https://www.google.com/");\n\
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);\n\
curl_exec($ch);\n\
$p = proc_open("echo test", [["pipe", "r"], ["pipe", "w"], ["pipe", "w"]], $pipes);\n\
echo stream_get_contents($pipes[1]);\n\
' >> test.php

CMD ["php", "test.php"]
```
2. Run the resulting Docker image.
3. Notice that it does not print `test` and hangs.

Commenting out the lines that install the Elastic APM Agent for PHP fixes the problem.

**Expected behavior**
Child processes spawned from PHP processes should not deadlock when the Elastic APM Agent for PHP is enabled.

### PR DESCRIPTION
**Pull request by intuibase**
Fixes #876
Fixes #962

### CODE REVIEW

**Comment by intuibase** [comment]
Fixed #876

**Comment by intuibase** [comment]
Fixes #962

**Review by SergeyKleyman — APPROVED** [review]
It's very interesting. I was right to be concerned about `pthread_atfork` - I made a mistake even just using it for logging :)
It seems `pthread_atfork` registration is inherited by child processes so it's redundant to re-register.

File: src/ext/platform_threads_linux.c [hunk 1]
```diff
@@ -524,7 +524,6 @@ static void callbackToLogForkAfterInParent()
 static void callbackToLogForkAfterInChild()
 {
     ELASTIC_APM_SIGNAL_SAFE_LOG_DEBUG( "After process fork (in child context); parent PID: %d", (int)getParentProcessId() );
-    registerCallbacksToLogFork();
 }
 
 void registerCallbacksToLogFork()
```
Comment (SergeyKleyman) [inline src/ext/platform_threads_linux.c:529]: Maybe it's worth adding a log statement before `pthread_atfork` something like:
```
ELASTIC_APM_SIGNAL_SAFE_LOG_DEBUG( "Registering callbacks to log process fork..." );
```

**Pull request merged**